app/models/concerns/filterable.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+# A module for types that are supposed to have FilterTaggings, calculated based
+# on their tags. Includes Taggable.
+module Filterable
+  extend ActiveSupport::Concern
+  include Taggable
+
+  included do
+    has_many :filter_taggings, as: :filterable, inverse_of: :filterable
+    has_many :filters, through: :filter_taggings
+
+    has_many :direct_filter_taggings,
+             -> { where(inherited: false) },
+             class_name: "FilterTagging",
+             as: :filterable
+    has_many :direct_filters,
+             source: :filter,
+             through: :direct_filter_taggings
+
+    after_destroy :clean_up_filter_taggings
+  end
+
+  # Update filters for this particular filterable.
+  def update_filters
+    FilterUpdater.new(self.class.base_class, [id], :main).update
+  end
+
+  # Destroy the filter-taggings and manually trigger callbacks.
+  def clean_up_filter_taggings
+    to_destroy = self.filter_taggings.to_a
+    to_destroy.each(&:destroy)
+    self.class.reindex_for_filter_changes([self.id], to_destroy, :main)
+  end
+
+  class_methods do
+    # Update the filters for all filterables in this relation.
+    def update_filters(async_update: false)
+      batch_size = ArchiveConfig.FILTER_UPDATE_BATCH_SIZE
+
+      select(:id).find_in_batches(batch_size: batch_size) do |batch|
+        updater = FilterUpdater.new(base_class, batch.map(&:id), :background)
+
+        if async_update
+          updater.async_update
+        else
+          updater.update
+        end
+
+        # Allow for progress messages in long-running updates.
+        yield if block_given?
+      end
+    end
+
+    # This is the callback that gets called when FilterUpdater is done with a
+    # single batch of Filterables of this type. Designed to reindex all of the
+    # Filterables whose filters changed. It's called in after_commit, to
```

Comment:
Because removing `taggings` triggers the removal of the corresponding `filter_taggings`, and `taggings` are deleted when the work is deleted, the `after_destroy :clean_up_filter_taggings` callback isn't strictly necessary. I think I may have written it during a period where I was trying to handle work deletion differently, and never realized that I could delete it once I abandoned the changes that I was working on.